pylearner.ts

```diff
@@ -1990,7 +2183,125 @@ function parseProofOutline(stmts: Statement[], i: number, precededByAssert: bool
+  let outerscopeVars : Immutable.Set<string> =  ImmutableSet();
+  const bindings = outerScope!.bindings;
+  for (var x in bindings)
+    outerscopeVars = outerscopeVars.add(x);
+  if(outerScope.outerScope != null)
+    return outerscopeVars.concat(getOuterScopeVariables(outerScope.outerScope));
+  return outerscopeVars;
+}
+
+function performAliasAnalysis(stmts: Statement[], i: number, previousStatementMaybeAliasRelation: MayAliasRelation) : MayAliasRelation {
+  if (!(stmts.length == i)) {
+    let stmt = stmts[i];
+    stmt.mayAliasRelation = previousStatementMaybeAliasRelation;
+    if(stmt instanceof ExpressionStatement) {
+      if(stmt.expr instanceof AssignmentExpression && stmt.expr.op == '=' && stmt.expr.lhs instanceof VariableExpression) {
+        if(stmt.expr.rhs instanceof VariableExpression) 
+          stmt.mayAliasRelation = stmt.mayAliasRelation.addMayAliases(stmt.expr.lhs.name,stmt.expr.rhs.name);
+        else stmt.mayAliasRelation = stmt.mayAliasRelation.removeVariableInMayAliasSets(stmt.expr.lhs.name);   
+      }
+    }
+    else if(stmt instanceof IfStatement) {
+        if(stmt.thenBody instanceof BlockStatement && stmt.elseBody instanceof BlockStatement &&  stmt.elseBody != null && stmt.thenBody != null) {
+        const stmtsThenBlock = stmt.thenBody.stmts;
+        const stmtsElseBlock = stmt.elseBody.stmts;
+        let mayAliasRelationThenBlock = performAliasAnalysis(stmtsThenBlock,0,stmt.mayAliasRelation)!;
+        let mayAliasRelationElseBlock = performAliasAnalysis(stmtsElseBlock,0,stmt.mayAliasRelation)!;
+        if(mayAliasRelationThenBlock&&mayAliasRelationElseBlock) {
+          const scopeThenBody = stmt.thenBody.scope!
+          const localVars = getLocalVariablesInScope(scopeThenBody);
+          const outerscopeVars = getOuterScopeVariables(scopeThenBody.outerScope!);
+          let varsToKeep = outerscopeVars.subtract(localVars);
+          const mayAliasRelationThenBlockWithVarsToKeep = mayAliasRelationThenBlock.keepVariablesInMayAliasSets(varsToKeep);
+          const mayAliasRelationElseBlockWithVarsToKeep = mayAliasRelationElseBlock.keepVariablesInMayAliasSets(varsToKeep);
+          let combinedMayAliasRelation = mayAliasRelationThenBlockWithVarsToKeep.combineWithMayAliasRelation(mayAliasRelationElseBlockWithVarsToKeep);
+          return performAliasAnalysis(stmts,i+1,combinedMayAliasRelation);
+        } 
+      }
+    }
+    else if(stmt instanceof WhileStatement){
+      if (stmt.body instanceof BlockStatement && stmt.body != null){
+        const stmtsBodyBlock = stmt.body.stmts;
+        let mayAliasRelationWhileBlock = performAliasAnalysis(stmtsBodyBlock,0,stmt.mayAliasRelation);
+        let mayAliasRelationWhileBlockCheck = performAliasAnalysis(stmtsBodyBlock,0,mayAliasRelationWhileBlock);
+        while (mayAliasRelationWhileBlock.mayAliasSets.toJS().toString() != mayAliasRelationWhileBlockCheck.mayAliasSets.toJS().toString()) {
```

Comment:
Add an `equals` method to class `MayAliasRelation` and use it here. Implement this `equals` method using [`equals`](https://immutable-js.com/docs/v4.3.5/Set/#equals()).